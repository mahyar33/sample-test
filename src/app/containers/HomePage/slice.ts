import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { TodosParams, TodosState } from './types';

export const initialState: TodosState = {
  todos: {
    loading: true,
    error: '',
    data: [],
  },
};

const todosSlice = createSlice({
  name: 'home',
  initialState,
  reducers: {
    setTodosList(state, action: PayloadAction<TodosState['todos']>) {
      const { payload } = action;
      state.todos.loading = payload.loading;
      state.todos.error = payload.error;
      state.todos.data = payload.data;
    },
    getTodosList(state, action: PayloadAction<TodosParams>) {
      state.todos.loading = true;
      state.todos.error = '';
      state.todos.data = [];
    },
  },
});

export const { actions, reducer, name: sliceKey } = todosSlice;
