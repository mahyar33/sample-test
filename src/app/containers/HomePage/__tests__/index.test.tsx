import React from 'react';
import { Store } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { HelmetProvider } from 'react-helmet-async';
import { configureAppStore } from 'store/configureStore';
import { actions, initialState } from '../slice';
import { HomePage } from '..';
import { i18n, translations } from 'locales/i18n';
import userEvent from '@testing-library/user-event';
import { createRenderer } from 'react-test-renderer/shallow';
const shallowRenderer = createRenderer();

function* mockHomePageSaga() {}

jest.mock('../saga', () => ({
  todosSaga: mockHomePageSaga,
}));

const renderHomePage = (store: Store) =>
  render(
    <Provider store={store}>
      <HelmetProvider>
        <HomePage />
      </HelmetProvider>
    </Provider>,
  );

describe('<renderHomePage />', () => {
  let store: ReturnType<typeof configureAppStore>;
  let component: ReturnType<typeof renderHomePage>;

  beforeEach(() => {
    jest.useFakeTimers();
    store = configureAppStore();
    component = renderHomePage(store);
    store.dispatch(actions.getTodosList({}));
    expect(store.getState().home).toEqual(initialState);
  });
  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
    component.unmount();
  });
  it('should render and match the snapshot', () => {
    shallowRenderer.render(
      <Provider store={store}>
        <HomePage />
      </Provider>,
    );
    const renderedOutput = shallowRenderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
  it('should dispatch action on mount', () => {
    store.dispatch(
      actions.setTodosList({ loading: false, error: '', data: [] }),
    );
    component.unmount();
    component = renderHomePage(store);
    expect(initialState.todos.data.length).toEqual(0);
    jest.runAllTimers();
    expect(store.getState().home.todos.loading).toBe(true);
  });
  it('should display loading indicator when state is loading', () => {
    component.unmount();
    component = renderHomePage(store);
    expect(component.container.querySelector('circle')).toBeInTheDocument();
  });
  it('should display empty table', async () => {
    const t = await i18n;
    store.dispatch(
      actions.setTodosList({ loading: false, error: '', data: [] }),
    );
    expect(
      component.queryByText(t(translations.noResult)!),
    ).toBeInTheDocument();
  });
  it('should display table', () => {
    store.dispatch(
      actions.setTodosList({
        loading: false,
        error: '',
        data: [{ userId: 1, id: 1, title: 'hi', completed: true }],
      }),
    );
    expect(component.queryByText('hi')).toBeInTheDocument();
  });
  it('should dispatch action on input change', () => {
    store.dispatch(
      actions.setTodosList({ loading: false, error: '', data: [] }),
    );
    const input = component.container.querySelector('input');
    userEvent.type(input!, 'Hello, World!');
    jest.runAllTimers();
    expect(store.getState().home.todos.loading).toBe(true);
  });
  it('should dispatch action on select change', () => {
    store.dispatch(
      actions.setTodosList({ loading: false, error: '', data: [] }),
    );
    const select = component.container.querySelector('select');
    userEvent.selectOptions(select!, 'true');
    jest.runAllTimers();
    expect(store.getState().home.todos.loading).toBe(true);
  });
});
