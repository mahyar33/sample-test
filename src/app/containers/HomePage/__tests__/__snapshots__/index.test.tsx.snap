// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<renderHomePage /> should render and match the snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "createReducer": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "injectedReducers": Object {
          "home": [Function],
        },
        "injectedSagas": Object {
          "home": Object {
            "mode": "@@saga-injector/daemon",
            "saga": [Function],
            "task": Object {
              "@@redux-saga/TASK": true,
              "cancel": [Function],
              "cont": [Function],
              "context": Object {},
              "end": [Function],
              "error": [Function],
              "id": 1,
              "isAborted": [Function],
              "isCancelled": [Function],
              "isRoot": true,
              "isRunning": [Function],
              "joiners": null,
              "meta": Object {
                "location": undefined,
                "name": "mockHomePageSaga",
              },
              "queue": Object {
                "abort": [Function],
                "addTask": [Function],
                "cancelAll": [Function],
                "getTasks": [Function],
              },
              "result": [Function],
              "setContext": [Function],
              "toPromise": [Function],
            },
          },
        },
        "replaceReducer": [Function],
        "runSaga": [Function],
        "subscribe": [Function],
        Symbol(Symbol.observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "createReducer": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "injectedReducers": Object {
            "home": [Function],
          },
          "injectedSagas": Object {
            "home": Object {
              "mode": "@@saga-injector/daemon",
              "saga": [Function],
              "task": Object {
                "@@redux-saga/TASK": true,
                "cancel": [Function],
                "cont": [Function],
                "context": Object {},
                "end": [Function],
                "error": [Function],
                "id": 1,
                "isAborted": [Function],
                "isCancelled": [Function],
                "isRoot": true,
                "isRunning": [Function],
                "joiners": null,
                "meta": Object {
                  "location": undefined,
                  "name": "mockHomePageSaga",
                },
                "queue": Object {
                  "abort": [Function],
                  "addTask": [Function],
                  "cancelAll": [Function],
                  "getTasks": [Function],
                },
                "result": [Function],
                "setContext": [Function],
                "toPromise": [Function],
              },
            },
          },
          "replaceReducer": [Function],
          "runSaga": [Function],
          "subscribe": [Function],
          Symbol(Symbol.observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <HomePage />
</Context.Provider>
`;
